firewood/src/merkle/node/branch.rs

```diff
@@ -21,6 +28,29 @@ pub struct BranchNode {
     pub(crate) children_encoded: [Option<Vec<u8>>; MAX_CHILDREN],
 }
 
+enum BranchDataLength {
+    None,
+    Length(u32),
+}
+
```

Comment:
This is confusing; why is this not an `Option<u32>`, or even better, `Option<NonZeroU32>`?

If zero is a valid value, consider https://crates.io/crates/nonmax

This way you get niche optimization with None so the size is smaller and serialization is easier.

Reply:
Would it be less confusing if it was 
```rust
enum BranchDataLength {
    NoData,
    /// `0` is a valid length, *empty* data and *no* data are treated separately
    Length(u32),
}
```
[`Shepmaster`](https://stackoverflow.com/users/155423/shepmaster) put this pretty nicely [in a comment on SO](https://stackoverflow.com/questions/30414068/can-i-use-the-null-pointer-optimization-for-my-own-non-pointer-types#comments-30414145):
>I'll just have to do more profiling to see if it's truly worth it. Using less bytes seems like an obvious win; using less bytes **and** mandatory math everywhere is less sure-fire.

(emphasis mine)

In any case, I didn't end up using this code. I'll delete it.

Reply:
deleted https://github.com/ava-labs/firewood/pull/373/commits/313cae7852d5ea66af1b43452634a4639bd1dfec